### PR TITLE
Yield-earning Warp Routes with ERC4626

### PR DESCRIPTION
**Pull request by ltyu**
### Description
Integrates ERC4626 compatible vaults. This PR assumes that the vault `asset` is the same as `wrappedToken`
- `_transferFromSender()` deposits into the vault
- `_transferTo()` withdraws from the vault
- `sweep()` redeems excess shares for the vault asset
- Uses internal `assetDeposited` store to keep track of assets deposited. Used to calculate excess shares to be withdrawn by owner
- Makes minor changes to existing test files to allow new tests to inherit

### Related issues

- Implements #2450

### Backward compatibility
Yes

### Testing
Manual/Unit Tests

### CODE REVIEW

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol:25]:
```suggestion
    constructor(
        ERC4626 _vault,
        address _mailbox
    ) HypERC20Collateral(_vault.asset(), _mailbox) {
        vault = _vault;
    }
```

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol:44]: should we do inifinite approvals here?

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol:15]: can we reuse `vault.balanceOf(address(this))` so we dont make this contract stateful?

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: this can hopefully be removed

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: we could skip the `withdraw` to `address(this)` and withdraw directly to `recipient` from `transferTo`?

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: please address this

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: please address this

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol:76]: this does quite a lot of redundant work
I think it's fine for this function to revert if `maxRedeem - convertToShares` underflows
is there a way to write this contract s.t. depositors get pro rata shares instead of it going to the owner?

Comment (yorhodes) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol:76]: I see now that we prescribed the owner being the recipient on the issue

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: resolved. following CEI

Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: resolved. following CEI

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol:15]: As per discussion. Will keep `assetDeposited`

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: As per discussion. Will keep assetDeposited

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: As per discussion, owner receives pro rata instead of each depositor. Removed `_excessVaultShares()`

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: Done

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (ltyu) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: Updated

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: I dont see any change here?

Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: this also violates the "checks, effects, interactions" pattern
https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [inline solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]:
```suggestion
        return vault.maxRedeem(address(this)) - vault.convertToShares(assetDeposited);
    }
```

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (yorhodes) [inline solidity/test/token/HypERC20CollateralVaultDeposit.t.sol]: why is this wrapped twice

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (yorhodes) [inline solidity/test/token/HypERC20CollateralVaultDeposit.t.sol:1]: can we reuse more internals from HypTokenTest here?
eg `_processTransfers`

File: solidity/contracts/token/HypERC20CollateralVaultDeposit.sol
Comment (yorhodes) [reply solidity/contracts/token/HypERC20CollateralVaultDeposit.sol]: updated or decided not do to this?

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (ltyu) [reply solidity/test/token/HypERC20CollateralVaultDeposit.t.sol]: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/d3d7ad7f973e52b60e9119cc98e82bca9a53f707/solidity/test/token/HypERC20.t.sol#L177-L178

Something similar is done in this test. If i understand correctly, it is used to set the localToken in the child class that is a `TokenRouter` type.

**Comment by avious00** [comment]
hey @ltyu still working on this?

**Comment by ltyu** [comment]
> hey @ltyu still working on this?

This is complete. Waiting on team approval

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (ltyu) [reply solidity/test/token/HypERC20CollateralVaultDeposit.t.sol:1]: As responded in TG:

"Any examples of where you want me to use more internals? I've mostly relied on _performRemoteTransfer()"

File: solidity/contracts/hooks/libs/AbstractMessageIdAuthHook.sol
Comment (yorhodes) [inline solidity/contracts/hooks/libs/AbstractMessageIdAuthHook.sol:82]: thanks for moving this in

File: solidity/test/isms/OPStackIsm.t.sol
Comment (yorhodes) [inline solidity/test/isms/OPStackIsm.t.sol:1]: nice

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (yorhodes) [inline solidity/test/token/HypERC20CollateralVaultDeposit.t.sol:85]: make a helper for this? or move to setup?

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (yorhodes) [inline solidity/test/token/HypERC20CollateralVaultDeposit.t.sol:115]: see https://book.getfoundry.sh/reference/forge-std/bound?highlight=bound#bound

**Review by yorhodes — COMMENTED** [review]
nice work!

**Review by yorhodes — APPROVED** [review]
nice work!

**Review by yorhodes — APPROVED** [review]
nice work!

**Review by yorhodes — APPROVED** [review]
nice work!

**Review by yorhodes — APPROVED** [review]
nice work!

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (ltyu) [reply solidity/test/token/HypERC20CollateralVaultDeposit.t.sol]: Updated

File: solidity/test/token/HypERC20CollateralVaultDeposit.t.sol
Comment (ltyu) [reply solidity/test/token/HypERC20CollateralVaultDeposit.t.sol]: Moved this and other duplicate calls to their own helpers

**Comment by avious00** [comment]
@yorhodes could you DM lee about the steps to merge

**Pull request merged**
